### ISSUE TITLE
Internal function tracing with callbacks failing due to private function

### ISSUE THREAD

**Issue by NickStallman** [DataDog/dd-trace-php#149]
This is slightly related to issue #113

It looks like there is a problem with tracing a function that uses callbacks, if that callback is private.
Specifically I am using a PHP S3 class "tpyo/amazon-s3-php-class" which is using the CURLOPT_HEADERFUNCTION feature.
curl_exec will execute the HEADERFUNCTION callback during data transmission.

As the callback is a private function of the S3 class, adding dd_trace to curl_exec throws these errors:

Message: Invalid callback S3Request::__responseHeaderCallback, cannot access private method S3Request::__responseHeaderCallback()
Message: curl_exec(): Could not call the CURLOPT_HEADERFUNCTION

So dd_trace is changing the scope and function visibility when curl_exec is run.

Curiously it doesn't appear to be all callbacks. I tried creating the below test script but it seems to function correctly.

```php
class a
{
        function __construct()
        {
                call_user_func(array($this, 'test'));
        }

        private function test()
        {
                echo "Hello World\n";
        }
}

dd_trace('call_user_func', function ($callback) {
        call_user_func($callback);
});

new a();
```

**Comment by labbati** [comment]
@pawelchcki any thoughts on this?

**Comment by pawelchcki** [comment]
Hey @NickStallman sorry for late response. Were you by any chance able to reproduce it in minimal setup? 
So far I wasn't able to reproduce it myself.

Is it the same PHP version and modules set as described in https://github.com/DataDog/dd-trace-php/issues/113 ?

**Comment by NickStallman** [comment]
Yep exactly the same configuration with tracer datadog-php-tracer_0.4.2-beta_amd64.

Here is a short test script that reliably produces the error.
Without dd_trace it executes perfectly, with dd_trace it fails with the error below:

```
Called curl_exec
PHP Warning:  Invalid callback a::headers, cannot access private method a::headers() in /root/test2.php on line 21

Warning: Invalid callback a::headers, cannot access private method a::headers() in /root/test2.php on line 21
PHP Warning:  curl_exec(): Could not call the CURLOPT_HEADERFUNCTION in /root/test2.php on line 21

Warning: curl_exec(): Could not call the CURLOPT_HEADERFUNCTION in /root/test2.php on line 21
Finished curl_exec
```

```php
<?php

class a
{
        function __construct()
        {
                $ch = curl_init('https://www.datadog.com/');
                curl_setopt($ch, CURLOPT_HEADERFUNCTION, array($this, 'headers'));
                $data = curl_exec($ch);
        }

        private function headers($ch, $headers)
        {
                var_dump($headers);
                return strlen($headers);
        }
}

dd_trace('curl_exec', function ($ch) {
        echo "Called curl_exec\n";
        curl_exec($ch);
        echo "Finished curl_exec\n";
});

new a();
```

**Comment by pawelchcki** [comment]
Thanks @NickStallman - this is very helpful! I'll post an update here once we have a fix for this/know more!

**Comment by NickStallman** [comment]
Has there been any progress on this issue? I notice #224 has the same issue repeated in an independent example.

This is a blocker for us to instrument one of our primary performance critical applications.

**Comment by labbati** [comment]
Hi @NickStallman, so far no progress on this topic and it revealed to be harder than expected!
Let us get back to you as we try to squeeze this into our current work.
Sorry again for the issue you are facing.

**Comment by Omicron7** [comment]
I just ran into this issue as well using a library called the phpCAS.
```
[16-Apr-2019 15:11:37 UTC] PHP Warning:  Invalid callback CAS_Request_CurlRequest::_curlReadHeaders, cannot access private method CAS_Request_CurlRequest::_curlReadHeaders() in /opt/datadog-php/dd-trace-sources/src/DDTrace/Integrations/Curl/CurlIntegration.php on line 55
[16-Apr-2019 15:11:37 UTC] PHP Warning:  curl_exec(): Could not call the CURLOPT_HEADERFUNCTION in /opt/datadog-php/dd-trace-sources/src/DDTrace/Integrations/Curl/CurlIntegration.php on line 55
```

Adding `curl` to the `DD_INTEGRATIONS_DISABLED` environment variable bypasses the issue for now.

**Comment by labbati** [comment]
Thanks for reporting this. This is related to how we pass context to our wrapper. We are actively working on this and his sibling (calls to `parent::xxx` from within a wrapped function). A fix for both (starting from the `parent::`) is expected in the next few days.

**Comment by SammyK** [comment]
Hey @NickStallman & @Omicron7! Thanks again for reporting this issue and sorry for the wait for a fix. PR's #284 & #303 should get rolled out into a release next week which should fix this issue. We'll let you know when the release is available! :)

**Comment by ctrlrsf** [comment]
@SammyK That's great to hear. This is also affecting an internal library we use so we've disabled curl integration for now.

Error file: /opt/datadog-php/dd-trace-sources/src/DDTrace/Integrations/Curl/CurlIntegration.php
Error line: 55
Error string: `curl_exec(): Could not call the CURLOPT_HEADERFUNCTION`

**Comment by SammyK** [comment]
Hey @NickStallman, @Omicron7, @ctrlrsf, et al! We just [released `0.21.0`](https://github.com/DataDog/dd-trace-php/releases) of the PHP tracer which should fix this issue. I'll go ahead and close this issue, but please let us know if you're still having and issue after upgrading and we can reopen. :)

**Comment by ctrlrsf** [comment]
Thanks, @SammyK. Testing now and will let you know if we have any issues.
